libs/math/include/math/tensor_reduce.hpp

```diff
@@ -18,39 +18,48 @@
 //------------------------------------------------------------------------------
 
 #include "math/base_types.hpp"
+#include "math/tensor_declaration.hpp"
 #include "math/tensor_slice_iterator.hpp"
 
 #include <cassert>
 
 namespace fetch {
 namespace math {
 
-// need to forward declare
-template <typename T, typename C>
-class Tensor;
-
+/**
+ * Applies given function along given axis on N-1 sized array
```

Comment:
it doesn't do this ON N-1 sized array right? you mean "resulting in a N-1 sized array" don't you?